File: packages/api/app/services/tokens.py
from __future__ import annotations

import hashlib
import hmac
import re
import secrets

_TOKEN_BYTES = 32
_USERNAME_RE = re.compile(r"[^a-z0-9_.-]+")


def normalize_username(username: str) -> str:
    normalized = _USERNAME_RE.sub("-", username.strip().lower()).strip("-._")
    if not normalized:
        raise ValueError("username is required")
    return normalized


def generate_token() -> str:
    return secrets.token_urlsafe(_TOKEN_BYTES)


def hash_token(token: str) -> str:
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def verify_token(token: str, token_hash: str) -> bool:
    return hmac.compare_digest(hash_token(token), token_hash)


def redact_secret(value: str) -> str:
    if not value:
        return "<empty>"
    if len(value) <= 8:
        return "<redacted>"
    return f"{value[:4]}…{value[-4:]}"

File: apps/api/src/services/tokens.ts
import { createHash, randomBytes, timingSafeEqual } from "node:crypto";

const TOKEN_BYTES = 32;
const INVALID_USERNAME_CHARACTERS = /[^a-z0-9_.-]+/g;
const EDGE_USERNAME_CHARACTERS = /^[-._]+|[-._]+$/g;

export function normalizeUsername(username: string): string {
  const normalized = username
    .trim()
    .toLowerCase()
    .replace(INVALID_USERNAME_CHARACTERS, "-")
    .replace(EDGE_USERNAME_CHARACTERS, "");

  if (!normalized) {
    throw new Error("username is required");
  }

  return normalized;
}

export function generateToken(): string {
  return randomBytes(TOKEN_BYTES).toString("base64url");
}

export function hashToken(token: string): string {
  return createHash("sha256").update(token, "utf8").digest("hex");
}

export function verifyToken(token: string, tokenHash: string): boolean {
  const actual = Buffer.from(hashToken(token), "utf8");
  const expected = Buffer.from(tokenHash, "utf8");

  return actual.length === expected.length && timingSafeEqual(actual, expected);
}

export function redactSecret(value: string): string {
  if (!value) {
    return "<empty>";
  }
  if (value.length <= 8) {
    return "<redacted>";
  }
  return `${value.slice(0, 4)}…${value.slice(-4)}`;
}
